### PR TITLE
module: zfs: avoid a taking a destroyed lock in zfs_zevent bits

### PR DESCRIPTION
**Pull request by kevans91**
### Motivation and Context

My laptop attempted to shutdown because of thermal issues, but in the process of shutting down it ended up panicking due to this upon unload of the zfs module. Unfortunately, this prevented the thermal situation from improving, but hopefully it'll have better luck next time. :-)

### Description

At shutdown time, we drain all of the zevents and set the
ZEVENT_SHUTDOWN flag.  In parallel, at least on FreeBSD, we may still be
winding down and call zfs_zevent_destroy() in the process.  For
instance, see: sysevent_worker() thread in
module/os/freebsd/spl/spl_sysevent.c -- we'll observe ESHUTDOWN then
back out.

There's no underlying zevent since we've drained zevent_list, so we can
just skip those bits and avoid leaking the zfs_zevent_t on the way out.

### How Has This Been Tested?

Loaded, unloaded the module

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by kevans91** [comment]
It's not immediately clear how this could have caused the ztest crash -- is this an intermittent/unrelated issue?

File: module/zfs/fm.c [hunk 1]
```diff
@@ -471,10 +471,19 @@ zfs_zevent_init(zfs_zevent_t **zep)
...
+	 * been drained and we're just cleaning up.  Avoid taking the
+	 * zevent_lock here because it may already be destroyed by the time we
+	 * are observing it.
+	 */
+	if ((zevent_flags & ZEVENT_SHUTDOWN) == 0) {
+		mutex_enter(&zevent_lock);
+		if (ze->ze_zevent)
+			list_remove(&ze->ze_zevent->ev_ze_list, ze);
+		mutex_exit(&zevent_lock);
+	}
```
Comment (behlendorf) [inline module/zfs/fm.c:486]: Perhaps this special handling should instead be placed in the `sysevent_worker()` thread around the `zfs_zevent_destroy()` call (or some other mechanism).  It looks to me like all other callers of the `zevent_lock` must be initialized.  Let's see if we can get @freqlabs or @amotin's thoughts on this.

Comment (kevans91) [reply module/zfs/fm.c:486]: No objection either way... we need the below kmem_free(), wasn't sure if we needed another new zfs_zevent_* KPI for just that (the conflation in this namespace of those operating on zevent_t vs. zfs_zevent_t is a bit confusing to me) or if the caller can just kmem_free directly instead.

Comment (freqlabs) [reply module/zfs/fm.c:486]: I am thinking `kmem_free` in `sysevent_worker`, with a comment explaining the race we need to avoid.

**Review by freqlabs — APPROVED** [review]

**Review by behlendorf — APPROVED** [review]
That works for me.

**Pull request merged**